
# Overview over Mjema_et_al-2025/Comparative analysis

## Procedure

To validate the predictions of the machine learning models, datasets of RNA-seq experiments that focused on the response of Arabidopsis thaliana to cold and heat treatments were used (see Comparative analysis/publications). No mutants were considered. All datasets were processed with the nf-core pipeline RNA-seq (version 3.18.0) to quantify the expression of the transcripts. The Arabidopsis thaliana TAIR10 reference genome was used. DESeq2 (version 1.49.3) was used to perform the differential gene expression analysis, where genes with p < 0.05 and Log2foldChange > 1 or < -1, where considered as differently expressed. 

The DEGs from the datasets from the publications were then compared with those predicted (Chapter 4), and a hypergeometric test was used to evaluate the statistical significance of these overlaps.

The DEGs from the datasets from the publications were also compared to the experimentally determined DEGs (Chapter 3) and a hypergeometric test was used to evaluate the statistical significance of these overlaps.

All analyses were performed with **R version 4.5.0**

## Structure of Comparative analysis

In "For chapter 3" and "For chapter 4" you can find a svg file with plots. Those plots show, how many percent of the genes (see **data**) were found in 0, 1, 2, 3, 4 or 5 of the publications. 

Here, you only see the structure of Comparative analysis/For chapter 3/cold-exp. All the other cold-exp and heat-exp folders from the "For chapter 3" and "For chapter 4" folder are structured equally. All files, that have the same name, but appear in different folders, are exactly the same.

![alt text](image.png)

**data**:
- file with the genes that were used to compare the differentially expressed genes (DEGs) from the publications against them
- in Chapter 4 those are genes, that were identified with machine learning (genes_winter.csv and genes_spring.csv)
- in Chapter 3 those are experimentally determined DEGs (DOWN_regulated.csv and UP_regulated.csv)
    
- samplesheets for the nf-core/rnaseq pipeline and DESeq2 

**resulting data**:

- lists_behind_heatmap_down.xls contains lists with genes (the ones from the file in **data** see above) that were found in 0, 1, 2, 3, 4, 5 or at least 1 of the publications
- down_table.csv is a extension of the file with genes from **data**, that contains the information in which publication(s) each gene was found

DESeq2:

- PCA plots from the DESeq2 analyses from all publications
- res_tables_winter.xlsx contains all the results from the DESeq2 analysis for all publications that examined cold-treatment (in the heat-exp folders you will find res_tables_spring.xlsx)
- signif_all_cold.xlsx contains all DEGs from the analyzed publications (in heat-exp it is signif_all_heat.xlsx)

rnaseq:

- multiqc reports for all publications
- star_salmon folder for each publication, that contains the quant.sf files and the tx2gene.tsv generated by the nf-core/rnaseq pipeline (not visible on the picture)

**results**:

- the overlap of genes between each publication and the genes from **data** was calculated and plotted
- in the plots you can also see the results of the hypergeometric tests

**scripts**:

- scripts for DESeq2 and nf-core/rnaseq for all publications that were examined
- script for the plots and hypergemetric tests in **results**
- results_DEGs.R contains the script to create the files (lists_behind_heatmap_down.xls and down_table.csv) in **resulting data**
